src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java

```diff
@@ -617,6 +622,59 @@ public T execute(final String merchantAccount, final PaymentData paymentData, fi
         }
     }
 
+    private PaymentTransactionInfoPlugin hppCompleteAuthWithPendingPayment(final UUID accountId,
+                                                                           final UUID kbPaymentId,
+                                                                           final UUID kbTransactionId,
+                                                                           final Iterable<PluginProperty> properties,
+                                                                           final CallContext context) throws PaymentPluginApiException {
+        final HppCompletedResult hppCompletedResult = AdyenPaymentServiceProviderHostedPaymentPagePort.parseAndVerifyRequestIntegrity(PluginProperties.toStringMap(properties), getConfigProperties(context),logger);
+        final PaymentServiceProviderResult authResult = PaymentServiceProviderResult.getPaymentResultForId(hppCompletedResult.getAuthResult());
+        AdyenResponsesRecord adyenResponsesRecord;
+        if(needCancelHppPayment(authResult)) {
+            cancelHppPayment(accountId, kbPaymentId, kbTransactionId, context);
+            final Iterable<PluginProperty> propertyWithErrorFromHppStatus = PluginProperties.merge(ImmutableMap.of(PROPERTY_FROM_HPP_TRANSACTION_STATUS,
+                                                                                                                   PaymentPluginStatus.ERROR),
+                                                                                                   properties);
+            adyenResponsesRecord = updateResponseWithAdditionalProperties(kbTransactionId,
+                                                                          authResult,
+                                                                          propertyWithErrorFromHppStatus,
+                                                                          context.getTenantId());
+        } else {
+            adyenResponsesRecord = updateResponseWithAdditionalProperties(kbTransactionId,
+                                                                          null,
+                                                                          properties,
+                                                                          context.getTenantId());
+        }
+        return buildPaymentTransactionInfoPlugin(adyenResponsesRecord);
+    }
+
+    private boolean needCancelHppPayment(final PaymentServiceProviderResult authResult) {
+        //We only cancel the hpp payment during a complete request if the authResult is either canceled or refused.
+        //In all other cases, we keep the status in pending until the confirmation of a notification.
+        return ImmutableList.of(PaymentServiceProviderResult.CANCELLED,
```

Comment:
This list (or set?) could be materialized (e.g. class constant), instead of rebuilding it all the time.

Reply:
Updated